ts/sbd-server/src/signal.ts

```diff
@@ -0,0 +1,407 @@
+import { DurableObject } from 'cloudflare:workers';
+
+import * as common from './common.ts';
+
+import { DoRateLimit } from './rate-limit.ts';
+import {
+  Msg,
+  MsgLbrt,
+  MsgLidl,
+  MsgAreq,
+  MsgAres,
+  MsgSrdy,
+  MsgKeep,
+  MsgNone,
+  MsgForward,
+} from './msg.ts';
+
+/**
+ * "SIGNAL" durable object.
+ */
+export class DoSignal extends DurableObject {
+  ctx: DurableObjectState;
+  env: Env;
+  queue: { [index: string]: Array<Uint8Array> };
+  alarmLock: boolean;
+  curLimit: number;
+  active: number;
+  lastCoord: number;
+  activeBytesReceived: number;
+  status: number;
+
+  constructor(ctx: DurableObjectState, env: Env) {
+    super(ctx, env);
+    this.ctx = ctx;
+    this.env = env;
+    this.queue = {};
+    this.alarmLock = false;
+    this.curLimit = 0;
+    this.active = common.timestamp();
+    this.lastCoord = common.timestamp();
+    this.activeBytesReceived = 0;
+    this.status = 200;
+  }
+
+  /**
+   * Client websockets are connected to a durable object identified
+   * by their own pubKey. When they send forward messages, those must
+   * be sent to durable objects identified by the destination pubKey.
+   * This is the api for those messages to be forwarded.
+   */
+  async forward(messageList: Array<Uint8Array>) {
+    if (this.status !== 200) {
+      return;
+    }
+
+    // MAYBE: buffer messages until handshake complete?
+    //        might not be needed since clients shouldn't publish the address
+    //        until handshake is complete
+
+    // This loop not technically needed, since our fetch ensures there
+    // is only ever one websocket connected.
+    for (const ws of this.ctx.getWebSockets()) {
+      for (const message of messageList) {
+        try {
+          ws.send(message);
+        } catch (e: any) {
+          console.error('forward error', e);
+          this.status = 500;
+          return;
+        }
+      }
+    }
+  }
+
+  /**
+   * This is the http endpoint for the "SIGNAL" durable object.
+   * The worker http fetch endpoint above forwards the request here.
+   * This function performs some checks, then upgrades the connection
+   * to a websocket.
+   */
+  async fetch(request: Request): Promise<Response> {
+    if (this.status !== 200) {
+      return new Response(JSON.stringify({ err: 'dead' }), {
+        status: this.status,
+      });
+    }
```

Comment:
added these checks that short-circuit cpu time when the object has already errored / closed